### PR TITLE
Pin glob expansion cache to EvaluationContext

### PR DESCRIPTION
**Pull request by cdmihai**
- Introduce an IFileSystemAbstraction. Not currently used.
- Pin glob expansion cache in EvaluationContext

Future work:
- Route all evaluation IO through the interface
- Add caching behind the interface
- Experiment with native vs managed implementations

### CODE REVIEW

**Comment by cdmihai** [comment]
Observations:
- Big improvements for 2nd run of large new style projects.
- Slight regression for 2nd run of large old style projects. No idea why this happens, have to investigate. Since there's no caching benefit, maybe the regression is just the caching overhead.
- Slight memory regression. This is expected since we're adding in caches.

Project | projects\msb-master 1st run | projects\msb-master 2nd run | Improvement | projects\msb-mihai 1st run | projects\msb-mihai 2nd run | Improvement
-- | -- | -- | -- | -- | -- | --
DotnetConsoleProject | 121.2632 | 99.9894 | 17.54349217 | 122.1236 | 94.1143 | 22.93520663
DotnetMvcProject | 179.4573 | 177.753 | 0.949696669 | 178.2479 | 146.2916 | 17.92800925
DotnetWebProject | 159.5837 | 168.683 | -5.701898126 | 160.9102 | 147.1444 | 8.554957983
Generated_100_100_v150 | 1888.4508 | 1537.3198 | 18.59360064 | 1895.519 | 1534.4352 | 19.04933688
LargeP2POldCsproj | 1340.6882 | 1348.5684 | -0.587772757 | 1321.233 | 1437.8958 | -8.829843033
OrchardCore | 6499.0302 | 6279.5715 | 3.37679151 | 6350.0762 | 2446.2104 | 61.47746385
Picasso | 645.8982 | 644.1481 | 0.270956011 | 644.7604 | 643.1669 | 0.247146072
Roslyn | 7996.702 | 7685.7516 | 3.888483027 | 7724.0881 | 3753.7372 | 51.40219594
SmallP2PNewCsproj | 486.8973 | 486.4501 | 0.091846884 | 485.4966 | 436.8795 | 10.01389093
SmallP2POldCsproj | 163.5558 | 172.1315 | -5.243287001 | 164.9468 | 173.9996 | -5.48831502
WebLargeCore | 4571.4902 | 4056.4795 | 11.26570719 | 4049.6147 | 2720.957 | 32.80948432

### SerialEvaluationSharedContextSecondRun Time (ms) (projects\msb-master vs projects\msb-mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 100.5897 -> 94.2138 (-6.339%)
DotnetWebProject | :white_check_mark: | yes | 168.8911 -> 146.3975 (-13.318%)
DotnetMvcProject | :white_check_mark: | yes | 176.1103 -> 145.6556 (-17.293%)
Picasso | ::ok_hand: | no | 655.5339 -> 648.5801 (-1.061%)
SmallP2POldCsproj | ::ok_hand: | no | 172.3634 -> 171.3784 (-0.571%)
SmallP2PNewCsproj | :white_check_mark: | yes | 492.1947 -> 440.4312 (-10.517%)
LargeP2POldCsproj | :red_circle: | yes | 1356.7806 -> 1454.7199 (7.219%)
Generated_100_100_v150 | :red_circle: | yes | 1548.1315 -> 1575.1851 (1.748%)
OrchardCore | :white_check_mark: | yes | 6182.9264 -> 2512.5411 (-59.363%)
Roslyn | :white_check_mark: | yes | 7551.1111 -> 3805.2701 (-49.606%)
WebLargeCore | :white_check_mark: | yes | 4062.6495 -> 2780.0105 (-31.571%)

### SerialEvaluationSharedContextSecondRun Memory (bytes) (projects\msb-master vs projects\msb-mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 5641320 -> 5580248 (-1.083%)
DotnetWebProject | :white_check_mark: | yes | 4011007 -> 3449165 (-14.008%)
DotnetMvcProject | :white_check_mark: | yes | 4780521 -> 3927736 (-17.839%)
Picasso | :red_circle: | yes | 14812588 -> 14938848 (0.852%)
SmallP2POldCsproj | :red_circle: | yes | 3391737 -> 3473657 (2.415%)
SmallP2PNewCsproj | :red_circle: | yes | 8446797 -> 10382226 (22.913%)
LargeP2POldCsproj | :white_check_mark: | yes | 25158824 -> 22772674 (-9.484%)
Generated_100_100_v150 | :white_check_mark: | yes | 26696030 -> 26572940 (-0.461%)
OrchardCore | :white_check_mark: | yes | 58762766 -> 55309411 (-5.877%)
Roslyn | :white_check_mark: | yes | 88833929 -> 85581963 (-3.661%)
WebLargeCore | :white_check_mark: | yes | 60154827 -> 56153110 (-6.652%)

### SerialEvaluationSharedContext Time (ms) (projects\msb-master vs projects\msb-mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 121.6581 -> 122.2128 (0.456%)
DotnetWebProject | :red_circle: | yes | 159.9814 -> 161.1962 (0.759%)
DotnetMvcProject | :white_check_mark: | yes | 180.6449 -> 179.263 (-0.765%)
Picasso | :ok_hand: | no | 646.8911 -> 651.4974 (0.712%)
SmallP2POldCsproj | :red_circle: | yes | 164.0592 -> 164.8847 (0.503%)
SmallP2PNewCsproj | :white_check_mark: | yes | 492.5681 -> 484.658 (-1.606%)
LargeP2POldCsproj | :white_check_mark: | yes | 1341.1605 -> 1323.2661 (-1.334%)
Generated_100_100_v150 | ::ok_hand: | no | 1916.076 -> 1915.9057 (-0.009%)
OrchardCore | ::ok_hand: | no | 6392.0705 -> 6363.9541 (-0.44%)
Roslyn | :white_check_mark: | yes | 8101.6224 -> 7677.8858 (-5.23%)
WebLargeCore | ::ok_hand: | no | 4488.9862 -> 4303.3962 (-4.134%)

### SerialEvaluationSharedContext Memory (bytes) (projects\msb-master vs projects\msb-mihai)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 2903979 -> 2975084 (2.449%)
DotnetWebProject | :red_circle: | yes | 4831780 -> 4917397 (1.772%)
DotnetMvcProject | :red_circle: | yes | 5577935 -> 5660786 (1.485%)
Picasso | :red_circle: | yes | 10494730 -> 10655584 (1.533%)
SmallP2POldCsproj | :red_circle: | yes | 4314178 -> 4393410 (1.837%)
SmallP2PNewCsproj | :white_check_mark: | yes | 7248703 -> 7080590 (-2.319%)
LargeP2POldCsproj | :red_circle: | yes | 20740954 -> 21417958 (3.264%)
Generated_100_100_v150 | :red_circle: | yes | 27071350 -> 27121170 (0.184%)
OrchardCore | :ok_hand: | no | 58695658 -> 59222154 (0.897%)
Roslyn | :ok_hand: | no | 86557569 -> 86984997 (0.494%)
WebLargeCore | :red_circle: | yes | 58768485 -> 61019881 (3.831%)

**Comment by cdmihai** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please
@dotnet-bot test Ubuntu16.04 Build for CoreCLR please

File: src/Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs [hunk 1]
```diff
@@ -38,6 +38,7 @@
 using ReservedPropertyNames = Microsoft.Build.Internal.ReservedPropertyNames;
 using NodeLoggingContext = Microsoft.Build.BackEnd.Logging.NodeLoggingContext;
 using ProjectLoggingContext = Microsoft.Build.BackEnd.Logging.ProjectLoggingContext;
+using Task = System.Threading.Tasks.Task;
```
Comment (AndyGerlicher) [inline src/Build/BackEnd/Components/RequestBuilder/RequestBuilder.cs]: revert?

File: src/Shared/FileSystem/FileSystemFactory.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+{
+    /// <summary>
+    /// Factory for <see cref="IFileSystemAbstraction"/>
+    /// </summary>
+    internal static class FileSystemFactory
+    {
+        /// <nodoc/>
+        public static IFileSystemAbstraction GetFileSystem()
+        {
+            // The windows-specific file system is only available on Windows Vista or higher
```
Comment (AndyGerlicher) [inline src/Shared/FileSystem/FileSystemFactory.cs]: Why is this commented out?

Comment (cdmihai) [reply src/Shared/FileSystem/FileSystemFactory.cs]: Need some experimentation to prove it's worth turning on. I'll create an issue about it.

**Comment by AndyGerlicher** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR  please

**Comment by cdmihai** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR

**Pull request merged**